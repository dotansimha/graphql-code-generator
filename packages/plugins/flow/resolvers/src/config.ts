import { RawResolversConfig } from '@graphql-codegen/visitor-plugin-common';

/**
 * @description This plugin generates resolvers signature based on your `GraphQLSchema`.
 *
 * It generates types for your entire schema: types, input types, enum, interface, scalar and union.
 *
 * This plugin requires you to use `@graphql-codegen/flow` as well, because it depends on it's types.
 */
export interface FlowResolversPluginConfig extends RawResolversConfig {}
