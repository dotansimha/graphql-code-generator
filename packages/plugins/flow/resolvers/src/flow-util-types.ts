export const FLOW_REQUIRE_FIELDS_TYPE = `export type $RequireFields<Origin, Keys> = $Diff<Origin, Keys> & $ObjMapi<Keys, <Key>(k: Key) => $NonMaybeType<$ElementType<Origin, Key>>>;`;
