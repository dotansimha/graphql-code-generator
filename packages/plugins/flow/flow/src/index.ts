import { Types, PluginFunction } from '@graphql-codegen/plugin-helpers';
import { parse, printSchema, visit, GraphQLSchema } from 'graphql';
import { FlowVisitor } from './visitor';
import { FlowPluginConfig } from './config';

export * from './visitor';
export * from './flow-variables-to-object';

export const plugin: PluginFunction<FlowPluginConfig, Types.ComplexPluginOutput> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: FlowPluginConfig
) => {
  const header = `// @flow\n`;
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);
  const visitor = new FlowVisitor(schema, config);

  const visitorResult = visit(astNode, {
    leave: visitor,
  });

  return {
    prepend: [header, ...visitor.getEnumsImports()],
    content: [visitor.scalarsDefinition, ...visitorResult.definitions].join('\n'),
  };
};
