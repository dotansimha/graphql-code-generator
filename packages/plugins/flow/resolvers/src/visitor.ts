import {
  ListTypeNode,
  NamedTypeNode,
  NonNullTypeNode,
  GraphQLSchema,
  ScalarTypeDefinitionNode,
  InputValueDefinitionNode,
  EnumTypeDefinitionNode,
} from 'graphql';
import autoBind from 'auto-bind';
import {
  RawResolversConfig,
  indent,
  ParsedResolversConfig,
  BaseResolversVisitor,
  DeclarationBlock,
  DeclarationKind,
} from '@graphql-codegen/visitor-plugin-common';
import { FlowOperationVariablesToObject } from '@graphql-codegen/flow';
import { FLOW_REQUIRE_FIELDS_TYPE } from './flow-util-types';

export const ENUM_RESOLVERS_SIGNATURE =
  'export type EnumResolverSignature<T, AllowedValues = any> = $ObjMap<T, () => AllowedValues>;';

export interface ParsedFlorResolversConfig extends ParsedResolversConfig {}

export class FlowResolversVisitor extends BaseResolversVisitor<RawResolversConfig, ParsedFlorResolversConfig> {
  constructor(pluginConfig: RawResolversConfig, schema: GraphQLSchema) {
    super(pluginConfig, null, schema);
    autoBind(this);
    this.setVariablesTransformer(new FlowOperationVariablesToObject(this.scalars, this.convertName));
  }

  protected _getScalar(name: string): string {
    return `$ElementType<Scalars, '${name}'>`;
  }

  protected applyRequireFields(argsType: string, fields: InputValueDefinitionNode[]): string {
    this._globalDeclarations.add(FLOW_REQUIRE_FIELDS_TYPE);
    return `$RequireFields<${argsType}, { ${fields.map(f => `${f.name.value}: *`).join(', ')} }>`;
  }

  protected applyOptionalFields(argsType: string, fields: readonly InputValueDefinitionNode[]): string {
    return argsType;
  }

  protected buildMapperImport(source: string, types: { identifier: string; asDefault?: boolean }[]): string {
    if (types[0] && types[0].asDefault) {
      return `import type ${types[0].identifier} from '${source}';`;
    }

    return `import { ${types.map(t => `type ${t.identifier}`).join(', ')} } from '${source}';`;
  }

  protected formatRootResolver(schemaTypeName: string, resolverType: string, declarationKind: DeclarationKind): string {
    return `${schemaTypeName}?: ${resolverType}${resolverType.includes('<') ? '' : '<>'}${this.getPunctuation(
      declarationKind
    )}`;
  }

  protected transformParentGenericType(parentType: string): string {
    return `ParentType = ${parentType}`;
  }

  ListType(node: ListTypeNode): string {
    return `?${super.ListType(node)}`;
  }

  NamedType(node: NamedTypeNode): string {
    return `?${super.NamedType(node)}`;
  }

  NonNullType(node: NonNullTypeNode): string {
    const baseValue = super.NonNullType(node);

    if (baseValue.startsWith('?')) {
      return baseValue.substr(1);
    }

    return baseValue;
  }

  protected applyMaybe(str: string): string {
    return `?${str}`;
  }

  protected clearMaybe(str: string): string {
    if (str.startsWith('?')) {
      return str.substr(1);
    }

    return str;
  }

  protected getTypeToUse(name: string): string {
    const resolversType = this.convertName('ResolversTypes');

    return `$ElementType<${resolversType}, '${name}'>`;
  }

  protected getParentTypeToUse(name: string): string {
    const resolversType = this.convertName('ResolversParentTypes');

    return `$ElementType<${resolversType}, '${name}'>`;
  }

  protected replaceFieldsInType(
    typeName: string,
    relevantFields: { fieldName: string; replaceWithType: string }[]
  ): string {
    return `$Diff<${typeName}, { ${relevantFields
      .map(f => `${f.fieldName}: * `)
      .join(', ')} }> & { ${relevantFields.map(f => `${f.fieldName}: ${f.replaceWithType}`).join(', ')} }`;
  }

  ScalarTypeDefinition(node: ScalarTypeDefinitionNode): string {
    const nameAsString = (node.name as any) as string;
    const baseName = this.getTypeToUse(nameAsString);
    this._collectedResolvers[node.name as any] = 'GraphQLScalarType';

    return new DeclarationBlock({
      ...this._declarationBlockConfig,
      blockTransformer(block) {
        return block;
      },
    })
      .export()
      .asKind('type')
      .withName(
        this.convertName(node, {
          suffix: 'ScalarConfig',
        })
      )
      .withBlock([indent(`...GraphQLScalarTypeConfig<${baseName}, any>`), indent(`name: '${node.name}'`)].join(', \n'))
      .string;
  }

  protected getPunctuation(declarationKind: DeclarationKind): string {
    return declarationKind === 'type' ? ',' : ';';
  }

  protected buildEnumResolverContentBlock(node: EnumTypeDefinitionNode, mappedEnumType: string): string {
    const valuesMap = `{| ${(node.values || [])
      .map(v => `${(v.name as any) as string}${this.config.avoidOptionals ? '' : '?'}: *`)
      .join(', ')} |}`;

    this._globalDeclarations.add(ENUM_RESOLVERS_SIGNATURE);

    return `EnumResolverSignature<${valuesMap}, ${mappedEnumType}>`;
  }

  protected buildEnumResolversExplicitMappedValues(
    node: EnumTypeDefinitionNode,
    valuesMapping: { [valueName: string]: string | number }
  ): string {
    return `{| ${(node.values || [])
      .map(v => {
        const valueName = (v.name as any) as string;
        const mappedValue = valuesMapping[valueName];

        return `${valueName}: ${typeof mappedValue === 'number' ? mappedValue : `'${mappedValue}'`}`;
      })
      .join(', ')} |}`;
  }
}
