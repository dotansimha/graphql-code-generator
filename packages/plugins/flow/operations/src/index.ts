import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
import { visit, concatAST, GraphQLSchema, Kind, FragmentDefinitionNode } from 'graphql';
import { FlowDocumentsVisitor } from './visitor';
import { LoadedFragment, optimizeOperations } from '@graphql-codegen/visitor-plugin-common';
import { FlowDocumentsPluginConfig } from './config';

export const plugin: PluginFunction<FlowDocumentsPluginConfig> = (
  schema: GraphQLSchema,
  rawDocuments: Types.DocumentFile[],
  config: FlowDocumentsPluginConfig
) => {
  const documents = config.flattenGeneratedTypes
    ? optimizeOperations(schema, rawDocuments, { includeFragments: true })
    : rawDocuments;

  const prefix = config.preResolveTypes
    ? ''
    : `type $Pick<Origin: Object, Keys: Object> = $ObjMapi<Keys, <Key>(k: Key) => $ElementType<Origin, Key>>;\n`;

  const allAst = concatAST(documents.map(v => v.document));
  const includedFragments = allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION);

  const allFragments: LoadedFragment[] = [
    ...(includedFragments as FragmentDefinitionNode[]).map(fragmentDef => ({
      node: fragmentDef,
      name: fragmentDef.name.value,
      onType: fragmentDef.typeCondition.name.value,
      isExternal: false,
    })),
    ...(config.externalFragments || []),
  ];

  const visitorResult = visit(allAst, {
    leave: new FlowDocumentsVisitor(schema, config, allFragments),
  });

  return {
    prepend: ['// @flow \n'],
    content: [prefix, ...visitorResult.definitions].join('\n'),
  };
};
