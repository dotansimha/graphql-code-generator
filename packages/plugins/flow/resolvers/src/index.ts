import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { RawResolversConfig } from '@graphql-codegen/visitor-plugin-common';
import { Types, PluginFunction, addFederationReferencesToSchema } from '@graphql-codegen/plugin-helpers';
import { parse, printSchema, visit, GraphQLSchema } from 'graphql';
import { FlowResolversVisitor } from './visitor';

/**
 * @description This plugin generates resolvers signature based on your `GraphQLSchema`.
 *
 * It generates types for your entire schema: types, input types, enum, interface, scalar and union.
 *
 * This plugin requires you to use `@graphql-codegen/flow` as well, because it depends on it's types.
 */
export interface RawFlowResolversConfig extends RawResolversConfig {}

export const plugin: PluginFunction<RawFlowResolversConfig, Types.ComplexPluginOutput> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: RawFlowResolversConfig
) => {
  const imports = ['type GraphQLResolveInfo'];
  const showUnusedMappers = typeof config.showUnusedMappers === 'boolean' ? config.showUnusedMappers : true;

  const gqlImports = `import { ${imports.join(', ')} } from 'graphql';`;

  const transformedSchema = config.federation ? addFederationReferencesToSchema(schema) : schema;

  const printedSchema = config.federation
    ? printSchemaWithDirectives(transformedSchema)
    : printSchema(transformedSchema);
  const astNode = parse(printedSchema);
  const visitor = new FlowResolversVisitor(config, transformedSchema);
  const visitorResult = visit(astNode, { leave: visitor });

  const defsToInclude: string[] = [visitor.getResolverTypeWrapperSignature()];

  if (visitor.hasFederation()) {
    defsToInclude.push(`
    export type ReferenceResolver<TResult, TReference, TContext> = (
      reference: TReference,
      context: TContext,
      info: GraphQLResolveInfo
    ) => Promise<TResult> | TResult;
    `);

    defsToInclude.push(`export type RecursivePick<T, U> = T`);
  }

  const header = `export type Resolver<Result, Parent = {}, Context = {}, Args = {}> = (
  parent: Parent,
  args: Args,
  context: Context,
  info: GraphQLResolveInfo
) => Promise<Result> | Result;

export type SubscriptionSubscribeFn<Result, Parent, Context, Args> = (
  parent: Parent,
  args: Args,
  context: Context,
  info: GraphQLResolveInfo
) => AsyncIterator<Result> | Promise<AsyncIterator<Result>>;

export type SubscriptionResolveFn<Result, Parent, Context, Args> = (
  parent: Parent,
  args: Args,
  context: Context,
  info: GraphQLResolveInfo
) => Result | Promise<Result>;

export interface SubscriptionSubscriberObject<Result, Key: string, Parent, Context, Args> {
  subscribe: SubscriptionSubscribeFn<{ [key: Key]: Result }, Parent, Context, Args>;
  resolve?: SubscriptionResolveFn<Result, { [key: Key]: Result }, Context, Args>;
}

export interface SubscriptionResolverObject<Result, Parent, Context, Args> {
  subscribe: SubscriptionSubscribeFn<mixed, Parent, Context, Args>;
  resolve: SubscriptionResolveFn<Result, mixed, Context, Args>;
}

export type SubscriptionObject<Result, Key: string, Parent, Context, Args> =
  | SubscriptionSubscriberObject<Result, Key, Parent, Context, Args>
  | SubscriptionResolverObject<Result, Parent, Context, Args>;

export type SubscriptionResolver<Result, Key: string, Parent = {}, Context = {}, Args = {}> =
  | ((...args: Array<any>) => SubscriptionObject<Result, Key, Parent, Context, Args>)
  | SubscriptionObject<Result, Key, Parent, Context, Args>;

export type TypeResolveFn<Types, Parent = {}, Context = {}> = (
  parent: Parent,
  context: Context,
  info: GraphQLResolveInfo
) => ?Types | Promise<?Types>;

export type IsTypeOfResolverFn<T = {}, Context = {}> = (obj: T, context: Context, info: GraphQLResolveInfo) => boolean | Promise<boolean>;

export type NextResolverFn<T> = () => Promise<T>;

export type DirectiveResolverFn<Result = {}, Parent = {}, Args = {}, Context = {}> = (
  next: NextResolverFn<Result>,
  parent: Parent,
  args: Args,
  context: Context,
  info: GraphQLResolveInfo
) => Result | Promise<Result>;

${defsToInclude.join('\n')}
`;

  const resolversTypeMapping = visitor.buildResolversTypes();
  const resolversParentTypeMapping = visitor.buildResolversParentTypes();
  const { getRootResolver, getAllDirectiveResolvers, mappersImports, unusedMappers, hasScalars } = visitor;

  if (hasScalars()) {
    imports.push('type GraphQLScalarTypeConfig');
  }

  if (showUnusedMappers && unusedMappers.length) {
    // eslint-disable-next-line no-console
    console.warn(`Unused mappers: ${unusedMappers.join(',')}`);
  }

  return {
    prepend: [gqlImports, ...mappersImports, ...visitor.globalDeclarations],
    content: [
      header,
      resolversTypeMapping,
      resolversParentTypeMapping,
      ...visitorResult.definitions.filter(d => typeof d === 'string'),
      getRootResolver(),
      getAllDirectiveResolvers(),
    ].join('\n'),
  };
};
