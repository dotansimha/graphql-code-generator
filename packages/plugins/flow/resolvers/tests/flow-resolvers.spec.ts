import '@graphql-codegen/testing';
import { buildSchema } from 'graphql';
import { plugin } from '../src';
import { schema } from '../../../typescript/resolvers/tests/common';
import { Types, mergeOutputs } from '@graphql-codegen/plugin-helpers';
import { ENUM_RESOLVERS_SIGNATURE } from '../src/visitor';

describe('Flow Resolvers Plugin', () => {
  describe('Enums', () => {
    it('Should not generate enum internal values resolvers when enum doesnt have enumValues set', async () => {
      const testSchema = buildSchema(/* GraphQL */ `
        type Query {
          v: MyEnum
        }

        enum MyEnum {
          A
          B
          C
        }
      `);
      const config = {};
      const result = await plugin(testSchema, [], config, { outputFile: '' });

      expect(result.prepend).not.toContain(ENUM_RESOLVERS_SIGNATURE);
      expect(result.content).not.toContain('EnumResolverSignature');
    });

    it('Should generate enum internal values resolvers when enum has enumValues set as object with explicit values', async () => {
      const testSchema = buildSchema(/* GraphQL */ `
        type Query {
          v: MyEnum
        }

        enum MyEnum {
          A
          B
          C
        }
      `);
      const config = {
        enumValues: {
          MyEnum: {
            A: 'val_1',
            B: 'val_2',
            C: 'val_3',
          },
        },
      };
      const result = await plugin(testSchema, [], config, { outputFile: '' });
      expect(result.prepend).not.toContain(ENUM_RESOLVERS_SIGNATURE);
      expect(result.content).not.toContain('EnumResolverSignature');
      expect(result.content).toContain(`export type MyEnumResolvers = {| A: 'val_1', B: 'val_2', C: 'val_3' |};`);
    });

    it('Should generate enum internal values resolvers when enum has enumValues set as external enum', async () => {
      const testSchema = buildSchema(/* GraphQL */ `
        type Query {
          v: MyEnum
        }

        enum MyEnum {
          A
          B
          C
        }
      `);
      const config = {
        enumValues: {
          MyEnum: 'MyCustomEnum',
        },
      };
      const result = await plugin(testSchema, [], config, { outputFile: '' });

      expect(result.prepend).toContain(ENUM_RESOLVERS_SIGNATURE);
      expect(result.content).toContain('EnumResolverSignature');
      expect(result.content).toContain(
        `export type MyEnumResolvers = EnumResolverSignature<{| A?: *, B?: *, C?: * |}, $ElementType<ResolversTypes, 'MyEnum'>>;`
      );
    });

    it('Should generate enum internal values resolvers when enum has mappers pointing to external enum', async () => {
      const testSchema = buildSchema(/* GraphQL */ `
        type Query {
          v: MyEnum
        }

        enum MyEnum {
          A
          B
          C
        }
      `);
      const config = {
        mappers: {
          MyEnum: 'MyCustomEnum',
        },
      };
      const result = await plugin(testSchema, [], config, { outputFile: '' });

      expect(result.prepend).toContain(ENUM_RESOLVERS_SIGNATURE);
      expect(result.content).toContain('EnumResolverSignature');
      expect(result.content).toContain(
        `export type MyEnumResolvers = EnumResolverSignature<{| A?: *, B?: *, C?: * |}, $ElementType<ResolversTypes, 'MyEnum'>>;`
      );
    });

    it('Should generate enum internal values resolvers when enum has enumValues set on a global level of all enums', async () => {
      const testSchema = buildSchema(/* GraphQL */ `
        type Query {
          v: MyEnum
        }

        enum MyEnum {
          A
          B
          C
        }
      `);
      const config = {
        enumValues: './enums',
      };
      const result = await plugin(testSchema, [], config, { outputFile: '' });

      expect(result.prepend).toContain(ENUM_RESOLVERS_SIGNATURE);
      expect(result.content).toContain('EnumResolverSignature');
      expect(result.content).toContain(
        `export type MyEnumResolvers = EnumResolverSignature<{| A?: *, B?: *, C?: * |}, $ElementType<ResolversTypes, 'MyEnum'>>;`
      );
    });
  });

  it('Should generate basic type resolvers', () => {
    const result = plugin(schema, [], {}, { outputFile: '' });

    expect(result).toMatchSnapshot();
  });

  it('Default values of args and compatibility with flow plugin', async () => {
    const testSchema = buildSchema(/* GraphQL */ `
      type Query {
        something(arg: String = "default_value"): String
      }
    `);

    const config: any = { noSchemaStitching: true };
    const result = await plugin(testSchema, [], config, { outputFile: '' });

    expect(result.prepend).toContain(
      `export type $RequireFields<Origin, Keys> = $Diff<Origin, Keys> & $ObjMapi<Keys, <Key>(k: Key) => $NonMaybeType<$ElementType<Origin, Key>>>;`
    );
    expect(result.content).toContain(
      `something?: Resolver<?$ElementType<ResolversTypes, 'String'>, ParentType, ContextType, $RequireFields<QuerySomethingArgs, { arg: * }>>,`
    );
  });

  it('Should generate ResolversParentTypes', () => {
    const result = plugin(schema, [], {}, { outputFile: '' }) as Types.ComplexPluginOutput;

    expect(result.content).toBeSimilarStringTo(`
      /** Mapping between all available schema types and the resolvers parents */
      export type ResolversParentTypes = {
        MyType: MyType,
        String: $ElementType<Scalars, 'String'>,
        MyOtherType: MyOtherType,
        Query: {},
        Subscription: {},
        Node: $ElementType<ResolversParentTypes, 'SomeNode'>,
        ID: $ElementType<Scalars, 'ID'>,
        SomeNode: SomeNode,
        MyUnion: $ElementType<ResolversParentTypes, 'MyType'> | $ElementType<ResolversParentTypes, 'MyOtherType'>,
        MyScalar: $ElementType<Scalars, 'MyScalar'>,
        Int: $ElementType<Scalars, 'Int'>,
        Boolean: $ElementType<Scalars, 'Boolean'>,
      };
    `);
  });

  it('Should generate the correct imports when schema has scalars', () => {
    const result = plugin(buildSchema(`scalar MyScalar`), [], {}, { outputFile: '' }) as Types.ComplexPluginOutput;

    expect(result.prepend).toContain(`import { type GraphQLResolveInfo } from 'graphql';`);
  });

  it('Should generate the correct imports when schema has no scalars', () => {
    const result = plugin(
      buildSchema(`type MyType { f: String }`),
      [],
      {},
      { outputFile: '' }
    ) as Types.ComplexPluginOutput;

    expect(result.prepend).not.toContain(
      `import { type GraphQLResolveInfo, type GraphQLScalarTypeConfig } from 'graphql';`
    );
  });

  it('Should generate valid output with args', async () => {
    const testSchema = buildSchema(/* GraphQL */ `
      type Query {
        release: String
      }
      type Mutation {
        random(byteLength: Int!): String!
      }
      schema {
        query: Query
        mutation: Mutation
      }
    `);
    const config = {
      contextType: 'GraphQLContext',
      skipTypename: true,
      addUnderscoreToArgsType: true,
    };
    const result = await plugin(testSchema, [], config, { outputFile: '' });
    const o = mergeOutputs([result]);
    expect(o).toContain(`$RequireFields<Mutation_RandomArgs, { byteLength: * }>>,`);
    expect(o).toContain(
      `export type $RequireFields<Origin, Keys> = $Diff<Origin, Keys> & $ObjMapi<Keys, <Key>(k: Key) => $NonMaybeType<$ElementType<Origin, Key>>>;`
    );
  });

  it('Should generate the correct resolver args type names when typesPrefix is specified', () => {
    const result = plugin(
      buildSchema(`type MyType { f(a: String): String }`),
      [],
      { typesPrefix: 'T' },
      { outputFile: '' }
    ) as Types.ComplexPluginOutput;

    expect(result.content).toBeSimilarStringTo(
      `f?: Resolver<?$ElementType<TResolversTypes, 'String'>, ParentType, ContextType, TMyTypeFArgs>,`
    );
  });

  it('should use MaybePromise in ResolverTypeWrapper', async () => {
    const testSchema = buildSchema(/* GraphQL */ `
      type MyQuery {
        posts: [Post]
      }

      type Post {
        author: String
        comment: String
      }

      schema {
        query: MyQuery
      }
    `);
    const content = (await plugin(
      testSchema,
      [],
      {
        rootValueType: 'MyRoot',
        asyncResolverTypes: true,
      } as any,
      { outputFile: 'graphql.ts' }
    )) as Types.ComplexPluginOutput;

    expect(content.content).toBeSimilarStringTo(`
      export type ResolverTypeWrapper<T> = Promise<T> | T;
    `);
  });
});
