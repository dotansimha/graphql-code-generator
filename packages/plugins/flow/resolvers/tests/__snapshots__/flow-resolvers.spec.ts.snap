// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow Resolvers Plugin Should generate basic type resolvers 1`] = `
Object {
  "content": "export type Resolver<Result, Parent = {}, Context = {}, Args = {}> = (
  parent: Parent,
  args: Args,
  context: Context,
  info: GraphQLResolveInfo
) => Promise<Result> | Result;

export type SubscriptionSubscribeFn<Result, Parent, Context, Args> = (
  parent: Parent,
  args: Args,
  context: Context,
  info: GraphQLResolveInfo
) => AsyncIterator<Result> | Promise<AsyncIterator<Result>>;

export type SubscriptionResolveFn<Result, Parent, Context, Args> = (
  parent: Parent,
  args: Args,
  context: Context,
  info: GraphQLResolveInfo
) => Result | Promise<Result>;

export interface SubscriptionSubscriberObject<Result, Key: string, Parent, Context, Args> {
  subscribe: SubscriptionSubscribeFn<{ [key: Key]: Result }, Parent, Context, Args>;
  resolve?: SubscriptionResolveFn<Result, { [key: Key]: Result }, Context, Args>;
}

export interface SubscriptionResolverObject<Result, Parent, Context, Args> {
  subscribe: SubscriptionSubscribeFn<mixed, Parent, Context, Args>;
  resolve: SubscriptionResolveFn<Result, mixed, Context, Args>;
}

export type SubscriptionObject<Result, Key: string, Parent, Context, Args> =
  | SubscriptionSubscriberObject<Result, Key, Parent, Context, Args>
  | SubscriptionResolverObject<Result, Parent, Context, Args>;

export type SubscriptionResolver<Result, Key: string, Parent = {}, Context = {}, Args = {}> =
  | ((...args: Array<any>) => SubscriptionObject<Result, Key, Parent, Context, Args>)
  | SubscriptionObject<Result, Key, Parent, Context, Args>;

export type TypeResolveFn<Types, Parent = {}, Context = {}> = (
  parent: Parent,
  context: Context,
  info: GraphQLResolveInfo
) => ?Types | Promise<?Types>;

export type IsTypeOfResolverFn<T = {}> = (obj: T, info: GraphQLResolveInfo) => boolean | Promise<boolean>;

export type NextResolverFn<T> = () => Promise<T>;

export type DirectiveResolverFn<Result = {}, Parent = {}, Args = {}, Context = {}> = (
  next: NextResolverFn<Result>,
  parent: Parent,
  args: Args,
  context: Context,
  info: GraphQLResolveInfo
) => Result | Promise<Result>;

export type ResolverTypeWrapper<T> = Promise<T> | T;

/** Mapping between all available schema types and the resolvers types */
export type ResolversTypes = {
  MyType: ResolverTypeWrapper<MyType>,
  String: ResolverTypeWrapper<$ElementType<Scalars, 'String'>>,
  MyOtherType: ResolverTypeWrapper<MyOtherType>,
  Query: ResolverTypeWrapper<{}>,
  Subscription: ResolverTypeWrapper<{}>,
  Node: $ElementType<ResolversTypes, 'SomeNode'>,
  ID: ResolverTypeWrapper<$ElementType<Scalars, 'ID'>>,
  SomeNode: ResolverTypeWrapper<SomeNode>,
  MyUnion: $ElementType<ResolversTypes, 'MyType'> | $ElementType<ResolversTypes, 'MyOtherType'>,
  MyScalar: ResolverTypeWrapper<$ElementType<Scalars, 'MyScalar'>>,
  Int: ResolverTypeWrapper<$ElementType<Scalars, 'Int'>>,
  Boolean: ResolverTypeWrapper<$ElementType<Scalars, 'Boolean'>>,
};

/** Mapping between all available schema types and the resolvers parents */
export type ResolversParentTypes = {
  MyType: MyType,
  String: $ElementType<Scalars, 'String'>,
  MyOtherType: MyOtherType,
  Query: {},
  Subscription: {},
  Node: $ElementType<ResolversParentTypes, 'SomeNode'>,
  ID: $ElementType<Scalars, 'ID'>,
  SomeNode: SomeNode,
  MyUnion: $ElementType<ResolversParentTypes, 'MyType'> | $ElementType<ResolversParentTypes, 'MyOtherType'>,
  MyScalar: $ElementType<Scalars, 'MyScalar'>,
  Int: $ElementType<Scalars, 'Int'>,
  Boolean: $ElementType<Scalars, 'Boolean'>,
};

export type MyDirectiveDirectiveArgs = {   arg: $ElementType<Scalars, 'Int'>,
  arg2: $ElementType<Scalars, 'String'>,
  arg3: $ElementType<Scalars, 'Boolean'>, };

export type MyDirectiveDirectiveResolver<Result, Parent, ContextType = any, Args = MyDirectiveDirectiveArgs> = DirectiveResolverFn<Result, Parent, ContextType, Args>;

export type MyTypeResolvers<ContextType = any, ParentType = $ElementType<ResolversParentTypes, 'MyType'>> = {
  foo?: Resolver<$ElementType<ResolversTypes, 'String'>, ParentType, ContextType>,
  otherType?: Resolver<?$ElementType<ResolversTypes, 'MyOtherType'>, ParentType, ContextType>,
  withArgs?: Resolver<?$ElementType<ResolversTypes, 'String'>, ParentType, ContextType, $RequireFields<MyTypeWithArgsArgs, { arg2: * }>>,
  __isTypeOf?: IsTypeOfResolverFn<ParentType>,
};

export type MyOtherTypeResolvers<ContextType = any, ParentType = $ElementType<ResolversParentTypes, 'MyOtherType'>> = {
  bar?: Resolver<$ElementType<ResolversTypes, 'String'>, ParentType, ContextType>,
  __isTypeOf?: IsTypeOfResolverFn<ParentType>,
};

export type QueryResolvers<ContextType = any, ParentType = $ElementType<ResolversParentTypes, 'Query'>> = {
  something?: Resolver<$ElementType<ResolversTypes, 'MyType'>, ParentType, ContextType>,
};

export type SubscriptionResolvers<ContextType = any, ParentType = $ElementType<ResolversParentTypes, 'Subscription'>> = {
  somethingChanged?: SubscriptionResolver<?$ElementType<ResolversTypes, 'MyOtherType'>, \\"somethingChanged\\", ParentType, ContextType>,
};

export type NodeResolvers<ContextType = any, ParentType = $ElementType<ResolversParentTypes, 'Node'>> = {
  __resolveType: TypeResolveFn<'SomeNode', ParentType, ContextType>,
  id?: Resolver<$ElementType<ResolversTypes, 'ID'>, ParentType, ContextType>,
};

export type SomeNodeResolvers<ContextType = any, ParentType = $ElementType<ResolversParentTypes, 'SomeNode'>> = {
  id?: Resolver<$ElementType<ResolversTypes, 'ID'>, ParentType, ContextType>,
  __isTypeOf?: IsTypeOfResolverFn<ParentType>,
};

export type MyUnionResolvers<ContextType = any, ParentType = $ElementType<ResolversParentTypes, 'MyUnion'>> = {
  __resolveType: TypeResolveFn<'MyType' | 'MyOtherType', ParentType, ContextType>,
};

export type MyScalarScalarConfig = {
  ...GraphQLScalarTypeConfig<$ElementType<ResolversTypes, 'MyScalar'>, any>, 
  name: 'MyScalar'
};

export type Resolvers<ContextType = any> = {
  MyType?: MyTypeResolvers<ContextType>,
  MyOtherType?: MyOtherTypeResolvers<ContextType>,
  Query?: QueryResolvers<ContextType>,
  Subscription?: SubscriptionResolvers<ContextType>,
  Node?: NodeResolvers<>,
  SomeNode?: SomeNodeResolvers<ContextType>,
  MyUnion?: MyUnionResolvers<>,
  MyScalar?: GraphQLScalarType<>,
};


/**
 * @deprecated
 * Use \\"Resolvers\\" root object instead. If you wish to get \\"IResolvers\\", add \\"typesPrefix: I\\" to your config.
 */
export type IResolvers<ContextType = any> = Resolvers<ContextType>;
export type DirectiveResolvers<ContextType = any> = {
  myDirective?: MyDirectiveDirectiveResolver<any, any, ContextType>,
};


/**
 * @deprecated
 * Use \\"DirectiveResolvers\\" root object instead. If you wish to get \\"IDirectiveResolvers\\", add \\"typesPrefix: I\\" to your config.
 */
export type IDirectiveResolvers<ContextType = any> = DirectiveResolvers<ContextType>;",
  "prepend": Array [
    "import { type GraphQLResolveInfo } from 'graphql';",
    "export type $RequireFields<Origin, Keys> = $Diff<Origin, Keys> & $ObjMapi<Keys, <Key>(k: Key) => $NonMaybeType<$ElementType<Origin, Key>>>;",
  ],
}
`;
